File: src/termscope/library.py
"""The TermScope jargon library: extensive, reference-style definitions.

Short definitions (in data/terms_*.json) are what fits on a notification. The
library augments each term with a fuller explanation sourced from an online
dictionary (Wikipedia) and cached offline in data/library.json, plus a source URL
for the "Learn more" action. If a term has no enriched entry yet, the library
falls back to its bundled short definition so callers always get something useful.

Run `python run.py enrich-library` to (re)build the cache from online sources.
"""
from __future__ import annotations

import json
from pathlib import Path

from . import paths
from .dictionary import TermEntry


def library_path() -> Path:
    return paths.BUNDLED_DATA_DIR / "library.json"


class Library:
    def __init__(self, path: Path | None = None):
        self.path = Path(path) if path else library_path()
        self._data: dict[str, dict] = {}
        self.reload()

    def reload(self) -> None:
        try:
            self._data = json.loads(self.path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError):
            self._data = {}

    def __len__(self) -> int:
        return len(self._data)

    def has(self, term_id: str) -> bool:
        return term_id in self._data

    def extended(self, entry: TermEntry) -> str:
        """Full reference definition for a term, or the short one as a fallback."""
        rec = self._data.get(entry.id)
        if rec and rec.get("extended"):
            return rec["extended"]
        return entry.definition

    def source(self, entry: TermEntry) -> tuple[str | None, str | None]:
        """(source_name, url) for the term's online reference, if known."""
        rec = self._data.get(entry.id) or {}
        return rec.get("source"), rec.get("url")

    def coverage(self) -> int:
        """How many entries were enriched from an online source (not just fallback)."""
        return sum(1 for r in self._data.values() if r.get("source") and r.get("url"))

File: src-tauri/src/library.rs
//! The TermScope jargon library — extended reference definitions, ported from
//! `legacy/src/termscope/library.py`. Reads the bundled `data/library.json`
//! (keyed by stable term id), falling back to the short definition when a term
//! has no enriched entry yet.

use serde::Serialize;
use std::collections::HashMap;
use std::path::Path;

use crate::dictionary::TermEntry;

#[derive(Default)]
pub struct Library {
    data: HashMap<String, serde_json::Value>,
}

#[derive(Serialize)]
pub struct Detail {
    pub extended: String,
    pub source: Option<String>,
    pub url: Option<String>,
}

impl Library {
    pub fn load(path: &Path) -> Self {
        let data = std::fs::read_to_string(path)
            .ok()
            .and_then(|t| serde_json::from_str::<HashMap<String, serde_json::Value>>(&t).ok())
            .unwrap_or_default();
        Self { data }
    }

    /// Full reference detail for a term, or the short definition as a fallback.
    pub fn detail(&self, entry: &TermEntry) -> Detail {
        let rec = self.data.get(&entry.id);
        let extended = rec
            .and_then(|r| r.get("extended"))
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string())
            .unwrap_or_else(|| entry.definition.clone());
        let source = rec
            .and_then(|r| r.get("source"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());
        let url = rec
            .and_then(|r| r.get("url"))
            .and_then(|v| v.as_str())
            .map(|s| s.to_string());
        Detail { extended, source, url }
    }
}
